torch/csrc/dynamo/eval_frame.c

```diff
@@ -438,10 +601,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
     // If the hit cache entry is not the head of the linked list,
     // move it to the head
     if (prev != NULL) {
-        CacheEntry* extra = get_cache_entry(frame->f_code);
+        ExtraState* extra = get_extra_state(frame->f_code);
+        CacheEntry* old_cache_entry = extract_cache_entry(extra);
+
         prev->next = e->next;
-        e->next = extra;
-        set_cache_entry(frame->f_code, e);
+        e->next = old_cache_entry;
```

Comment:
Yup, here it is. This violates the contracts: you can't set old_cache_entry to an owning field, because you only got a borrowed reference from extract_cache_entry